src/box/lua/config/utils/odict.lua

```diff
@@ -0,0 +1,209 @@
+--  | print('od.c', od.c) -- 3
+--  |
+--  | for k, v in odict.pairs(od) do
+--  |     print(k, v)
+--  | end
+--  | -- print: a, 1
+--  | -- print: b, 2
+--  | -- print: c, 3
+--
+-- If an element is deleted and added again, it is added to the
+-- end.
+--
+--  | local od = odict.new()
+--  |
+--  | od.a = 1
+--  | od.b = 2
+--  | od.c = 3
+--  |
+--  | od.b = nil
+--  | od.b = 4
+--  |
+--  | for k, v in odict.pairs(od) do
+--  |     print(k, v)
+--  | end
+--  | -- print: a, 1
+--  | -- print: c, 3
+--  | -- print: b, 4
+--
+-- If an element is changed (without prior deletion), it remains
+-- on the same position.
+
+local REINDEX_THRESHOLD = 1000
+
+local registry = setmetatable({}, {__mode = 'k'})
+
+-- Generator function for odict.pairs().
+--
+-- Returns (key, value).
+local function gen(od, prev_key)
+    local ctx = registry[od]
+
+    -- The previous key is nil only on the first call of the
+    -- generator function.
+    local id = prev_key == nil and 0 or ctx.key2id[prev_key]
+    -- NB: This assert doesn't catch all the kinds of changes
+    -- during an iteration. It rather verifies a precondition
+    -- for the following cycle.
+    assert(id ~= nil, 'ordered dictionary is changed during iteration')
+
+    while id <= ctx.max_id do
+        id = id + 1
+        local key = ctx.id2key[id]
+        -- id2key may have holes for IDs <= max_id in place of
+        -- deleted items.
+        --
+        -- id2key may contain a stalled entry, because __newindex
+        -- is not called on assignment of an existing field,
+        -- including assignment to nil.
+        if key ~= nil and od[key] ~= nil then
+            return key, od[key]
+        end
+    end
+end
+
+-- Returns a Lua iterator triplet (gen, param, state).
+--
+-- Usage:
+--
+--  | for k, v in odict.pairs(od) do
+--  |     <...>
+--  | end
+--
+-- The iterator stability guarantees are the same as for usual
+-- pairs().
+--
+-- Quote from the Lua 5.1 reference manual:
+--
+-- > The behavior of `next` is undefined if, during the traversal,
+-- > you assign any value to a non-existent field in the table.
+-- > You may however modify existing fields. In particular, you
+-- > may clear existing fields.
+local function _pairs(od)
+    return gen, od, nil
+end
+
+-- Delete the key from the id<->key mappings.
+local function trash(ctx, key)
+    local id = ctx.key2id[key]
+    if id ~= nil then
+        ctx.key2id[key] = nil
+        ctx.id2key[id] = nil
+
+        -- No need to grow the IDs if the same key is assigned and
+        -- deleted in a loop.
+        --
+        -- for <...> do
+        --     od.x = 'x'
+        --     od.x = nil
+        -- end
+        if id == ctx.max_id then
+            ctx.max_id = ctx.max_id - 1
+        end
+    end
+end
+
+-- Track the new key in the id<->key mappings.
+local function track(ctx, key)
+    local id = ctx.max_id + 1
+    ctx.id2key[id] = key
+    ctx.key2id[key] = id
+    ctx.max_id = id
+end
```

Comment:
Nit:
This seems more like a `new_index()`/`add_key()` function.
Actually, it would be nice to have `drop_key()` and `add_key()` instead of `trash()` and `track()`.

Reply:
Answered [above](https://github.com/tarantool/tarantool/pull/9730#discussion_r1505606234).

Reply:
Updated, see the linked discussion thread.